h/form.py

```diff
@@ -89,6 +95,78 @@ def configure_environment(config):  # pragma: no cover
     config.registry[ENVIRONMENT_KEY] = create_environment(base)
 
 
+def handle_form_submission(request, form, on_success, on_failure):
+    """
+    Handle the submission of the given form in a standard way.
+
+    :param request: the Pyramid request
+
+    :param form: the form that was submitted
+    :type form: deform.form.Form
+
+    :param on_success:
+        A callback function to be called if the form validates successfully.
+
+        This function should carry out the action that the form submission
+        requests (for example for a change password form, this function would
+        change the user's password) and return the view callable result that
+        should be returned if this is not an XHR request.
+
+        If on_success() returns ``None`` then ``handle_form_submission()``
+        will return ``HTTPFound(location=request.url)`` by default.
+    :type on_success: callable
+
+    :param on_failure:
+        A callback function that will be called if form validation fails in
+        order to get the view callable result that should be returned if this is
+        not an XHR request.
+    :type on_failure: callable
+
+    """
+    try:
+        appstruct = form.validate(request.POST.items())
+    except deform.ValidationFailure:
+        result = on_failure()
+    else:
+        result = on_success(appstruct)
+
+        if result is None:
+            result = httpexceptions.HTTPFound(location=request.url)
+
+        request.session.flash(_("Success. We've saved your changes."),
```

Comment:
I'm not entirely sure this is going to turn out to be the right place for this, but I guess we can deal with that when it becomes a problem.